File: libraries/analytics_debugger/src/main/java/com/tokopedia/analyticsdebugger/debugger/data/source/ApplinkLogDBSource.java
package com.tokopedia.analyticsdebugger.debugger.data.source;

import android.content.Context;

import com.tokopedia.analyticsdebugger.database.ApplinkLogDB;
import com.tokopedia.analyticsdebugger.database.TkpdAnalyticsDatabase;
import com.tokopedia.analyticsdebugger.debugger.AnalyticsDebuggerConst;
import com.tokopedia.analyticsdebugger.debugger.domain.model.ApplinkLogModel;

import java.util.Date;
import java.util.HashMap;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;

public class ApplinkLogDBSource {

    private ApplinkLogDao applinkLogDao;

    @Inject
    public ApplinkLogDBSource(Context context) {
        applinkLogDao = TkpdAnalyticsDatabase.getInstance(context).applinkLogDao();
    }

    public Observable<Boolean> deleteAll() {
        return Observable.unsafeCreate(subscriber -> {
            applinkLogDao.deleteAll();
            subscriber.onNext(true);
        });
    }

    public Observable<Boolean> insertAll(ApplinkLogModel data) {
        return Observable.just(data).map(analyticsLogData -> {
            ApplinkLogDB applinkLogDB = new ApplinkLogDB();
            applinkLogDB.setApplink(data.getApplink());
            applinkLogDB.setTraces(data.getTraces());
            applinkLogDB.setTimestamp(new Date().getTime());

            applinkLogDao.insertAll(applinkLogDB);
            return true;
        });
    }

    public Observable<List<ApplinkLogDB>> getData(HashMap<String, Object> params) {
        return Observable.just(params).map(params1 -> {
            int page;

            if (params1.containsKey(AnalyticsDebuggerConst.PAGE))
                page = (int) params1.get(AnalyticsDebuggerConst.PAGE);
            else page = 0;

            String search = "%%";

            if (params1.containsKey(AnalyticsDebuggerConst.KEYWORD)) {
                String keyword = (String) params1.get(AnalyticsDebuggerConst.KEYWORD);
                search = "%" + keyword + "%";
            }

            int offset = 20 * page;
            return applinkLogDao.getData(search, offset);
        });
    }
}

File: libraries/analytics_debugger/src/main/java/com/tokopedia/analyticsdebugger/debugger/data/source/ApplinkLogDBSource.kt
package com.tokopedia.analyticsdebugger.debugger.data.source

import android.content.Context

import com.tokopedia.analyticsdebugger.database.ApplinkLogDB
import com.tokopedia.analyticsdebugger.database.TkpdAnalyticsDatabase
import com.tokopedia.analyticsdebugger.debugger.AnalyticsDebuggerConst
import com.tokopedia.analyticsdebugger.debugger.domain.model.ApplinkLogModel

import java.util.Date
import java.util.HashMap

import javax.inject.Inject

import rx.Observable

class ApplinkLogDBSource @Inject
constructor(context: Context) {

    private val applinkLogDao: ApplinkLogDao

    init {
        applinkLogDao = TkpdAnalyticsDatabase.getInstance(context).applinkLogDao()
    }

    fun deleteAll(): Observable<Boolean> {
        return Observable.unsafeCreate { subscriber ->
            applinkLogDao.deleteAll()
            subscriber.onNext(true)
        }
    }

    fun insertAll(data: ApplinkLogModel): Observable<Boolean> {
        return Observable.just(data).map { analyticsLogData ->
            val applinkLogDB = ApplinkLogDB()
            applinkLogDB.applink = data.applink
            applinkLogDB.traces = data.traces
            applinkLogDB.timestamp = Date().time

            applinkLogDao.insertAll(applinkLogDB)
            true
        }
    }

    fun getData(params: HashMap<String, Any>): Observable<List<ApplinkLogDB>> {
        return Observable.just(params).map { params1 ->
            val page: Int

            if (params1.containsKey(AnalyticsDebuggerConst.PAGE))
                page = params1[AnalyticsDebuggerConst.PAGE] as Int
            else
                page = 0

            var search = "%%"

            if (params1.containsKey(AnalyticsDebuggerConst.KEYWORD)) {
                val keyword = params1[AnalyticsDebuggerConst.KEYWORD] as String?
                search = "%$keyword%"
            }

            val offset = 20 * page
            applinkLogDao.getData(search, offset)
        }
    }
}
